File: packages/wire/transfer_chunker_test.go
package wire

import "testing"

// piece is a comparable snapshot of a FramePiece for table assertions.
type piece struct {
	block, off, plen int
	last             bool
	payload          string
}

// collect runs ReChunk over data delivered as the given chunk groups and records each piece.
func collect(t *testing.T, chunks [][]byte, block, frame int) []piece {
	t.Helper()
	stream := func(fn func([]byte) error) error {
		for _, c := range chunks {
			if err := fn(c); err != nil {
				return err
			}
		}
		return nil
	}
	var got []piece
	err := ReChunk(stream, block, frame, func(p FramePiece) error {
		got = append(got, piece{p.BlockIdx, p.FrameOff, len(p.Payload), p.LastInBlock, string(p.Payload)})
		return nil
	})
	if err != nil {
		t.Fatalf("ReChunk: %v", err)
	}
	return got
}

func seq(n int) []byte {
	b := make([]byte, n)
	for i := range b {
		b[i] = byte(i)
	}
	return b
}

func TestReChunkEmptyStream(t *testing.T) {
	if got := collect(t, nil, 8, 4); len(got) != 0 {
		t.Errorf("empty stream produced %d pieces, want 0", len(got))
	}
}

func TestReChunkSubFrameFile(t *testing.T) {
	got := collect(t, [][]byte{{1, 2, 3}}, 8, 4)
	want := []piece{{0, 0, 3, true, string([]byte{1, 2, 3})}}
	assertPieces(t, got, want)
}

func TestReChunkExactBlockMultiple(t *testing.T) {
	// block=8, frame=4, 16 bytes = 2 blocks × 2 frames, all full.
	got := collect(t, [][]byte{seq(16)}, 8, 4)
	wantShape := [][4]int{{0, 0, 4, 0}, {0, 4, 4, 1}, {1, 0, 4, 0}, {1, 4, 4, 1}}
	assertShape(t, got, wantShape)
	assertBytes(t, got, seq(16))
}

func TestReChunkAwkwardBoundaries(t *testing.T) {
	// Same 16 bytes delivered as 3+7+6 must produce output identical to one 16-byte chunk.
	d := seq(16)
	got := collect(t, [][]byte{d[0:3], d[3:10], d[10:16]}, 8, 4)
	wantShape := [][4]int{{0, 0, 4, 0}, {0, 4, 4, 1}, {1, 0, 4, 0}, {1, 4, 4, 1}}
	assertShape(t, got, wantShape)
	assertBytes(t, got, d)
}

func TestReChunkFinalPartialBlock(t *testing.T) {
	// block=8, frame=4, 10 bytes → block0 full (2 frames), block1 one 2-byte frame.
	got := collect(t, [][]byte{seq(10)}, 8, 4)
	assertShape(t, got, [][4]int{{0, 0, 4, 0}, {0, 4, 4, 1}, {1, 0, 2, 1}})
}

func TestReChunkFrameEqualsBlock(t *testing.T) {
	// block=8, frame=8, 20 bytes → 8, 8, 4; one frame per block, each the block's last.
	got := collect(t, [][]byte{seq(20)}, 8, 8)
	assertShape(t, got, [][4]int{{0, 0, 8, 1}, {1, 0, 8, 1}, {2, 0, 4, 1}})
}

func TestReChunkPropagatesEmitError(t *testing.T) {
	stream := func(fn func([]byte) error) error { return fn(seq(16)) }
	sentinel := NewTransferError(FailSinkError, "stop")
	calls := 0
	err := ReChunk(stream, 8, 4, func(FramePiece) error {
		calls++
		return sentinel
	})
	if err != sentinel {
		t.Errorf("ReChunk error = %v, want sentinel", err)
	}
	if calls != 1 {
		t.Errorf("kept emitting after error: %d calls, want 1", calls)
	}
}

func assertPieces(t *testing.T, got, want []piece) {
	t.Helper()
	if len(got) != len(want) {
		t.Fatalf("got %d pieces, want %d: %+v", len(got), len(want), got)
	}
	for i := range got {
		if got[i] != want[i] {
			t.Errorf("piece %d = %+v, want %+v", i, got[i], want[i])
		}
	}
}

// assertShape checks [blockIdx, frameOff, payloadLen, lastInBlock(0/1)] per piece.
func assertShape(t *testing.T, got []piece, want [][4]int) {
	t.Helper()
	if len(got) != len(want) {
		t.Fatalf("got %d pieces, want %d: %+v", len(got), len(want), got)
	}
	for i, w := range want {
		last := 0
		if got[i].last {
			last = 1
		}
		if got[i].block != w[0] || got[i].off != w[1] || got[i].plen != w[2] || last != w[3] {
			t.Errorf("piece %d = [%d %d %d %d], want %v",
				i, got[i].block, got[i].off, got[i].plen, last, w)
		}
	}
}

// assertBytes checks the concatenated payloads equal want.
func assertBytes(t *testing.T, got []piece, want []byte) {
	t.Helper()
	var all []byte
	for _, p := range got {
		all = append(all, []byte(p.payload)...)
	}
	if string(all) != string(want) {
		t.Errorf("reassembled bytes = %v, want %v", all, want)
	}
}

File: packages/protocol/src/transfer-chunker.test.ts
import { describe, it, expect } from 'vitest';
import { reChunk } from './transfer-chunker.js';

async function* fromChunks(chunks: number[][]): AsyncGenerator<Uint8Array> {
  for (const c of chunks) yield new Uint8Array(c);
}
async function collect(stream: AsyncIterable<Uint8Array>, block: number, frame: number) {
  const pieces: Array<{
    blockIdx: number;
    frameOff: number;
    payload: number[];
    lastInBlock: boolean;
  }> = [];
  for await (const p of reChunk(stream, block, frame)) {
    pieces.push({
      blockIdx: p.blockIdx,
      frameOff: p.frameOff,
      payload: [...p.payload],
      lastInBlock: p.lastInBlock,
    });
  }
  return pieces;
}

describe('reChunk', () => {
  it('yields nothing for a 0-byte stream', async () => {
    expect(await collect(fromChunks([]), 8, 4)).toEqual([]);
  });

  it('emits one small last frame for a sub-frame file', async () => {
    const pieces = await collect(fromChunks([[1, 2, 3]]), 8, 4);
    expect(pieces).toEqual([{ blockIdx: 0, frameOff: 0, payload: [1, 2, 3], lastInBlock: true }]);
  });

  it('splits an exact block-multiple across full frames, flagging block ends', async () => {
    // block=8, frame=4, 16 bytes = 2 blocks × 2 frames, all full.
    const data = Array.from({ length: 16 }, (_, i) => i);
    const pieces = await collect(fromChunks([data]), 8, 4);
    expect(pieces.map((p) => [p.blockIdx, p.frameOff, p.lastInBlock])).toEqual([
      [0, 0, false],
      [0, 4, true],
      [1, 0, false],
      [1, 4, true],
    ]);
    expect(pieces.flatMap((p) => p.payload)).toEqual(data);
  });

  it('re-chunks awkward input boundaries into aligned frames', async () => {
    // Same 16 bytes delivered as 3+7+6 — output must be identical to one 16-byte chunk.
    const data = Array.from({ length: 16 }, (_, i) => i);
    const pieces = await collect(
      fromChunks([
        [0, 1, 2],
        [3, 4, 5, 6, 7, 8, 9],
        [10, 11, 12, 13, 14, 15],
      ]),
      8,
      4,
    );
    expect(pieces.map((p) => [p.blockIdx, p.frameOff, p.lastInBlock])).toEqual([
      [0, 0, false],
      [0, 4, true],
      [1, 0, false],
      [1, 4, true],
    ]);
    expect(pieces.flatMap((p) => p.payload)).toEqual(data);
  });

  it('handles a final partial block (tail smaller than a block)', async () => {
    // block=8, frame=4, 10 bytes → block0 full (2 frames), block1 has one 2-byte frame.
    const data = Array.from({ length: 10 }, (_, i) => i);
    const pieces = await collect(fromChunks([data]), 8, 4);
    expect(pieces.map((p) => [p.blockIdx, p.frameOff, p.payload.length, p.lastInBlock])).toEqual([
      [0, 0, 4, false],
      [0, 4, 4, true],
      [1, 0, 2, true],
    ]);
  });

  it('frame equal to block yields one frame per block', async () => {
    const data = Array.from({ length: 20 }, (_, i) => i); // block=8: 8,8,4
    const pieces = await collect(fromChunks([data]), 8, 8);
    expect(pieces.map((p) => [p.blockIdx, p.frameOff, p.payload.length, p.lastInBlock])).toEqual([
      [0, 0, 8, true],
      [1, 0, 8, true],
      [2, 0, 4, true],
    ]);
  });
});
